### PR TITLE
Isort running synchronously via python on editor contents. 

### PR DESCRIPTION
**Pull request by arl-o**
Howdy Lexicalunit, it's me from [this conversation](https://github.com/bh/atom-python-isort/issues/34)! (With a new username.) 

I finally got everything together and 99% working. 

Changes:
- [x] All isort behavior is now called (synchronously) in python, on the contents of the editor. It runs on highlighted code if there is any, and supports sort-on-save and check-on-save. 
- [x] There are now a lot of supported [isort settings](https://github.com/timothycrosley/isort/wiki/isort-Settings) in the package configuration. 
- [x] Added an option to get rid of the annoying status-bar-check-mark-thing, and use built-in atom notifications instead. 
- [x] Added editor-language-context checks to the `check-imports` and `sort-imports` commands, so they'll only appear when the window is in a python context. 
- [x] Temporarily! added activation commands, so that errors in the onWillSave event could be avoided by reloading and not activating the package. 

Todos:
- [ ] Re-add support for isort path, would need to pass the path to python and add it to `sys.path`. 
- [ ] Possibly allow the status bar option to be toggled back and forth and take effect without restarts. (Not super high priority, I just couldn't figure out the scope.)
- [x] Maybe add some more [isort settings](https://github.com/timothycrosley/isort/wiki/isort-Settings). 
- [x] Make any syntax/cleanliness/less-hack-y changes you'd recommend. eg: I used mostly `underscore_case`, but the original package was in `camelCase`. Does that matter to you? 
- [x] I haven't done any checks for things like the windows python path, or to see if the python process works in python 3. I'll do the python 3 support soon, but I don't have any windows machines for those checks. 
- [x] Remove activation commands when finished. 

I realize that this is a large-ish re-write, so if you'd rather not do with the trouble of fixing what isn't broken, I totally understand.

### CODE REVIEW

**Comment by lexicalunit** [comment]
This is super awesome and I can't wait to test it out! Hopefully I'll have some free time this weekend 🤞

**Comment by lexicalunit** [comment]
I haven't forgot about this! Just taking a while to get to it.

File: lib/index.coffee [hunk 1]
```diff
@@ -1,57 +1,196 @@
...
-    @status = new StatusDialog pi
-    pi.setStatusDialog(@status)
+    @subs.add atom.config.observe 'atom-isort.showStatusBar', (value) ->
+      atom.workspace.observeTextEditors (editor) ->
+        if not value
+          if pi.statusDialog?
+            pi.removeStatusbarItem()
+        else
+          # TODO: re-adding status bar isn't working, not sure how to get it to.
+          # Scope in JS is confusing. Need to call consumeStatusBar?
```
Comment (lexicalunit) [inline lib/index.coffee]: Yeah this was originally written before the status bar was a consumable service. It doesn't surprise me that things don't quite work right when re-adding the status bar.

**Comment by arl-o** [comment]
Awesome, thank you for the edits! I'm totally new to coffeescript, so style defaulted to what I'm used to writing. 

Also somewhat new to github collaboration: I don't need to approve your edits, do I? Or am I supposed to now merge your `arl-o-master` branch with my `master` branch, then go from there? 

Still need some free time to finish the TODOs, probably tomorrow.

**Comment by lexicalunit** [comment]
Yep! You could merge my branch into yours to take my edits wholesale, then continue editing your PR. To update your PR you could use a force push to wholesale update the PR... or you could go the way of updating the PR by stacking commits on top of it, eschewing the use of force push. It's mostly down to personal preference.

**Comment by arl-o** [comment]
Your edits are merged, python 3.6.1 checks passed! 

Last couple TODOs are optional:

Do we need an isort path? Or will `import isort` in python cover 100% of odd use-cases? 

I'll add some more isort settings, (eg `force_to_top`), as soon as I get over this cold, and get home from short travelling.

**Comment by arl-o** [comment]
Okay, some final commits are in. I added some more options and removed the activation commands.

Only remaining quibble is that, without activation commands, it adds (on my machine) 42 ms to startup, which isn't great. However, without activation commands, we don't get 'sort on save' or 'check on save.' 

TODO's are all done, 'cept for `isort path`, see last comment.

**Comment by lexicalunit** [comment]
Oops, I just noticed that I have npm setup to use a work artifactory instead of the typical registry.npmjs.org. We might want to omit `package-lock.json`, or edit it to ensure that there's no `maven-devtools-repo-prod.mfpaws.com` junk in there.

**Comment by arl-o** [comment]
Ok, simplest solution is omit `package-lock.json`?

**Comment by lexicalunit** [comment]
Yep, definitely the simplest solution. Tho from what I gather from googling it, it seem like it's intended to be something that people check in to repositories, I guess? \shrug

**Comment by lexicalunit** [comment]
As for the `isort` path TODO: Typically any given Python will have it's own environment that it operates within. Be that the system environment or one of several virtual environments provided tools such as pyenv/virturalenv or conda. So it's not so much that you'd want to be able to configure the path to the isort module as it is that you're want to be able to configure your path the `python` binary.

**Comment by arl-o** [comment]
Awesome! Package lock is removed, sorry I was out for a couple days. I also double-checked the settings, everything looks good on that front.

**Comment by lexicalunit** [comment]
@arl-o I've been playing around with these changes locally and I noticed something that I'm not sure if it's just me. When you run `atom-isort` on a file, does it put your cursor at the end of the file?

For example, open `atom-isort.py`, move around the import statements, put your cursor on like line 10 or so, run the `Sort imports` command, then press the up arrow key. For me my cursor suddenly jumps to the end of the file.

I've dealt with an issue like this before and found a naive solution that works well enough IMHO: Just find the cursor location before you run the command, then run the command, then set the cursor to its previous location. It's not always correct because lines could have been removed by isort, but it generally does pretty well.

**Comment by arl-o** [comment]
I think it does reset the cursor, I'll try implementing your solution tonight and let you know.

**Comment by lexicalunit** [comment]
Actually, this delta resolves the issue for me:

```
diff --git a/lib/atom-isort.coffee b/lib/atom-isort.coffee
index b07f3aa..3bc5c59 100644
--- a/lib/atom-isort.coffee
+++ b/lib/atom-isort.coffee
@@ -113,7 +113,9 @@ class AtomIsort
       self.updateStatusbarText '⧗', true
       if response['new_contents'].length > 0
         if insert_type == 'set'
+          pos = editor.getCursorScreenPosition()
           editor.setText response['new_contents']
+          editor.setCursorScreenPosition(pos)
         else if insert_type == 'insert'
           editor.insertText response['new_contents']
         self.updateStatusbarText '√', true
```

I'm gonna merge this PR and then make this change. Excellent work on this PR btw!!

**Comment by lexicalunit** [comment]
@arl-o I just released v2.0.0 which includes your awesome improvements! 🎉 

Your changes also enabled me to add some spec tests, thanks again!

**Pull request merged**